### ISSUE TITLE
two questions about forward encoder (self, x) in ngswin. py

### ISSUE THREAD

**Issue by xxuantgo** [rami0205/NGramSwin#12]
Hello author! I found two areas of difficulty in handling and understanding while reading the code. Can you help me solve the confusion? Thank you! I presented my specific problem in the code section (forward encoder (self, x) in ngswin. py), please see Question1 and Question2.
    
        # Question1: why "self.inter_pool(c1, num_patches1)"?
        # in Figure(a) Overall architecture of NGswin,
        # it should receive c0(from shallow module)and e1_(from encoder_layer1)
      
        # Question2: why e3_[B, HW/(4^2), D],e3 [B, HW/(8^2), D]? there is no Patch Merging in encoder_layer3
        # e3_ and e3 should have the same shape

    def forward_encoder(self, x):
        """in ngswin.py"""
        _, _, H, W = x.size()
        x = self.shallow_extract(x)
        shallow_short = x
        # Across Stage Pooling Cascading
        c0, num_patches0 = rearrange(x, 'b c h w -> b (h w) c'), (H, W)  # [B, HW, 1D], (H, W). ASPC
        e1_, e1, num_patches1 = self.encoder_layer1(c0, num_patches0)  # [B, HW, D], [B, HW/(2^2), D], (H/2, W/2). encoder-stage1
        c1 = torch.cat([self.inter_pool(c0, num_patches0), e1], dim=-1)  # [B, HW/(2^2), 2D], (H, W). ASPC
        e2_, e2, num_patches2 = self.encoder_layer2(c1, num_patches1)  # [B, HW/(2^2), D], [B, HW/(4^2), D], (H/4, W/4). encoder-stage2

        # Question1: why "self.inter_pool(c1, num_patches1)"?
        # in Figure(a) Overall architecture of NGswin,
        # it should receive c0(from shallow module)and e1_(from encoder_layer1)
        c2 = torch.cat([self.inter_pool(c1, num_patches1), e2], dim=-1)  # [B, HW/(4^2), 3D], (H, W). ASPC
        e3_, e3, num_patches3 = self.encoder_layer3(c2, num_patches2)  # [B, HW/(4^2), D], [B, HW/(8^2), D], (H/8, W/8). encoder-stage3

        # Question2: why e3_[B, HW/(4^2), D],e3 [B, HW/(8^2), D]? there is no Patch Merging in encoder_layer3
        # e3_ and e3 should have the same shape
        num_patches_list = [num_patches0, num_patches1, num_patches2]  # (H, W), (H/2, W/2), (H/4, W/4)
        # bottleneck
        out, num_patches_scdp = self.bottleneck(shallow_short, [e1_, e2_, e3_], num_patches_list)  # [B, HW, D], (H, W)
        return out, num_patches_scdp, c0, e1_

<img width="448" alt="微信图片_20240301221412" src="https://github.com/rami0205/NGramSwin/assets/157219491/fa2d8631-896f-44de-b0d9-c14e42ef6e4b">
